src/Core/Optimise/Inline.hs

```diff
@@ -1,103 +1,84 @@
 {-# LANGUAGE ScopedTypeVariables #-}
 module Core.Optimise.Inline
-  ( inlineVariable
-  , betaReduce
-  , complexSafePropag
+  ( inlineVariablePass
   ) where
 
-import qualified Data.Map.Merge.Strict as Map
-import qualified Data.Map.Strict as Map
+import qualified Data.VarMap as VarMap
 import qualified Data.VarSet as VarSet
-import Data.Generics
+import qualified Data.Map.Strict as Map
+import Data.VarSet (IsVar(..))
 import Data.Triple
-import Data.Either
-import Data.Maybe
 
 import Core.Optimise
-import Syntax (Resolved)
-
-import Debug.Trace
-import Pretty (pretty)
 
 limit :: Int
 limit = 500
 
--- The "opposite" (dual?) of a propagation pass, inlining attempts to
--- "pull in" the definition of a variable it considers cheap.
-inlineVariable :: TransformPass
-inlineVariable = pass go where
-  go it@(CotRef v _) = do
-    def <- find v
-    case def of
-      Just term -> do
-        cost <- score term
-        if not (isLambda term) || cost >= limit || recursive term v
-           then pure it
-           else pure term
-      Nothing -> pure it
-  go x = pure x
+data InlineScope a = InlineScope { scores :: VarMap.Map (CoAtom a, Int)
+                                 -- , vars :: VarMap.Map (CoTerm a)
+                                 , cons :: VarSet.Set }
 
-  isLambda CotLam{} = True
-  isLambda _ = False
-
-betaReduce :: TransformPass
-betaReduce = pass go where
-  go term = case term of
-    CotApp (CotLam Small (var, tp) body) ex -> do
-      (var', ref) <- invent tp
-      pure . CotLet [(var', tp, ex)] .  substitute (Map.singleton var ref) $ body
-    CotTyApp (CotLam Big (var, _) body) tp ->
-      pure $ substituteInTys (Map.singleton var tp) body
-    _ -> pure term
+inlineVariablePass :: IsVar a => [CoStmt a] -> [CoStmt a]
```

Comment:
Is this still live?

Reply:
Yeah, this now works again. It's horribly slow in GHCi, but it works.